logstage/logstage-core/src/main/scala/izumi/logstage/api/Log.scala

```diff
@@ -157,7 +157,17 @@ object Log {
   final case class Message(template: StringContext, args: LogContext) {
     def argsMap: Map[String, Set[Any]] = args.map(kv => (kv.name, kv.value)).toMultimap
 
-    def ++(that: Message): Message = Message(StringContext(template.parts ++ that.template.parts: _*), args ++ that.args)
+    def ++(that: Message): Message = {
+      // this ain't pretty but avoids multiple iterations on the same parts collection compared to using .init/.tail
+      val (thisInit, thisTails) = template.parts.splitAt(template.parts.length - 1)
+      val thisTail = thisTails.head
```

Comment:
@an-tex Could you please make it work if the left Message is empty (`Message(StringContext(), Nil)`) or consists of one part (`Message("")`), and add a test for that?

Reply:
Is `StringContext()` a valid StringContext? Can't see how StringContext.checkLengths(args) could NOT throw an exception for any arguments. An empty Message is also defined as `def empty: Message = Message(new StringContext(""), Nil)`. The code already works for `Message.empty` and `Message("")` but I've added tests documenting it.

Reply:
@an-tex Yeah, `checkLengths` is not called in StringContext constructor, so nothing stops you from constructing a `Message(StringContext(), Nil)` manually - because it's a value that can legally happen I think it shouldn't throw in `++`. Or I guess you could add a call `template.checkLengths(args)` in Message constructor which would forbid such a value from appearing.

Reply:
Fair point. As throwing an exception in the `Message` constructor might break existing functionality for people already using an empty StringContext for whatever reason I've opted in to just handle  that case in the `++`

Reply:
Thanks!